include/rmm/mr/device/system_memory_resource.hpp

```diff
@@ -0,0 +1,187 @@
+ *  https://developer.nvidia.com/blog/nvidia-grace-hopper-superchip-architecture-in-depth/.
+ */
+class system_memory_resource final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a new system memory resource.
+   *
+   * By default if no parameters are specified, this memory resource pass through to malloc/free
+   * for allocation/deallocation.
+   *
+   * However, when GPU memory is over-subscribed, system allocated memory will migrate to the GPU
+   * and cannot migrate back, thus causing other CUDA calls to fail with out-of-memory errors. To
+   * work around this problem, we can reserve some GPU memory as headroom for other CUDA calls.
+   * Doing this check can be expensive, so only large buffer above the given threshold will be
+   * checked.
+   *
+   * @param headroom_size Size of the reserved GPU memory as headroom
+   * @param threshold_size Size of the allocation above which to check for headroom
+   */
+  explicit system_memory_resource(std::optional<std::size_t> headroom_size  = std::nullopt,
+                                  std::optional<std::size_t> threshold_size = std::nullopt)
+  {
+    RMM_EXPECTS(rmm::mr::detail::sam::is_supported(),
+                "System memory allocator is not supported with this hardware/software version.");
+    headroom_size_ = headroom_size.value_or(0);
+    if (headroom_size_ == 0) {
+      // If headroom is not specified, always pass through to malloc.
+      threshold_size_ = threshold_size.value_or(std::numeric_limits<std::size_t>::max());
+    } else {
+      // If headroom is specified, by default we check every allocation.
+      threshold_size_ = threshold_size.value_or(0);
+    }
+  }
+
+  ~system_memory_resource() override                    = default;
+  system_memory_resource(system_memory_resource const&) = default;  ///< @default_copy_constructor
+  system_memory_resource(system_memory_resource&&)      = default;  ///< @default_move_constructor
+  system_memory_resource& operator=(system_memory_resource const&) =
+    default;  ///< @default_copy_assignment{system_memory_resource}
+  system_memory_resource& operator=(system_memory_resource&&) =
+    default;  ///< @default_move_assignment{system_memory_resource}
+
+ private:
+  /**
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer will have at minimum 256 byte alignment.
+   *
+   * The stream argument is ignored.
+   *
+   * @param bytes The size of the allocation
+   * @param stream This argument is ignored
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, [[maybe_unused]] cuda_stream_view stream) override
+  {
+    auto const aligned{rmm::align_up(bytes, rmm::CUDA_ALLOCATION_ALIGNMENT)};
+    void* ptr{malloc(aligned)};
+
+    if (bytes >= threshold_size_) {
+      auto const free        = rmm::available_device_memory().first;
+      auto const allocatable = std::max(free - headroom_size_, 0UL);
```

Comment:
*issue*: this subtraction is between two unsigned integers, and therefore wraps whenever `free` is less than `headroom_size_`. So exactly when you want to _not_ allocate on the GPU, you will end up determining that you can allocate all of `aligned` bytes on the GPU.

Reply:
Good catch, fixed.